### PR TITLE
[WASI-NN] Support tensorflow lite backend

### PR DESCRIPTION
**Pull request by gusye1234**
support tensorflow lite backend for wasi-nn.

- [x] implementation
- [x] tests
- [x] CI
- [x] examples

### CODE REVIEW

File: utils/wasi-nn/install-pytorch.sh [hunk 1]
```diff
@@ -15,17 +15,17 @@ PYTORCH_SHA="b76d6dd4380e2233ce6f7654e672e13aae7c871231d223a4267ef018dcbfb616"
 
 for i in "$@"; do
   case $i in
-    --disable-cxx11-abi)
-      PYTORCH_LINK="libtorch"
-      PYTORCH_SHA="b5ddadc9addc054d8503f4086546f0cbcfdc3fc70087863bbd7b0e3300e3247f"
-      shift
-      ;;
+  --disable-cxx11-abi)
```
Comment (q82419) [inline utils/wasi-nn/install-pytorch.sh:18]: I think this file is not necessary to be modified?

File: utils/wasi-nn/download-tflite-fixtures.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+TODIR=$1
+
+FIXTURE_EX=https://raw.githubusercontent.com/second-state/wasm-learning/master/rust/birds_v1
```
Comment (q82419) [inline utils/wasi-nn/download-tflite-fixtures.sh:3]: Change to use the `WasmEdge-WASINN-examples` repo after [this PR](https://github.com/second-state/WasmEdge-WASINN-examples/pull/12) being merged.

**Review by q82419 — CHANGES_REQUESTED** [review]

File: utils/wasi-nn/install-pytorch.sh (hunk 1 above)
Comment (gusye1234) [reply utils/wasi-nn/install-pytorch.sh:18]: Yeah... It seems like the styling tool of my editor automatically modify it. It didn't involve any functional change.

File: plugins/wasi_nn/CMakeLists.txt [hunk 3]
```diff
@@ -43,24 +43,126 @@ install(TARGETS wasmedgePluginWasiNN DESTINATION ${CMAKE_INSTALL_LIBDIR}/wasmedg
...
+    )
+    FetchContent_GetProperties(wasmedge_tensorflow_deps)
+
+    if(NOT wasmedge_tensorflow_deps_POPULATED)
+      message(STATUS "Fetching WasmEdge-tensorflow-dep repository")
+      FetchContent_Populate(wasmedge_tensorflow_deps)
+      message(STATUS "Fetching WasmEdge-tensorflow-dep repository - done")
+    endif()
+
+    # -hardcode remove in future-------------------------------
```
Comment (q82419) [inline plugins/wasi_nn/CMakeLists.txt:140]: Please clean the comments.

File: plugins/wasi_nn/wasinnfunc.cpp [hunk 4]
```diff
@@ -296,6 +292,50 @@ Expect<uint32_t> WasiNNLoad::body(const Runtime::CallingFrame &Frame,
...
+    char *BinPtr = MemInst->getPointer<char *>(GraphBuilders[0], BinLen);
+    if (unlikely(BinPtr == nullptr)) {
+      spdlog::error("[WASI-NN] Failed when accessing the Weight memory.");
+      return static_cast<uint32_t>(WASINN::ErrNo::InvalidArgument);
+    }
+    // Add a new graph.
+    Env.NNGraph.emplace_back(static_cast<WASINN::Backend>(Encoding));
+    auto &Graph = Env.NNGraph.back();
+
+    // -------------------------
```
Comment (q82419) [inline plugins/wasi_nn/wasinnfunc.cpp:320]: Delete the meaningless comment.

File: plugins/wasi_nn/wasinnfunc.cpp [hunk 5]
```diff
@@ -296,6 +292,50 @@ Expect<uint32_t> WasiNNLoad::body(const Runtime::CallingFrame &Frame,
...
+      Env.NNGraph.pop_back();
+      return static_cast<uint32_t>(WASINN::ErrNo::InvalidArgument);
+    }
+    Graph.TFLiteOps = TfLiteInterpreterOptionsCreate();
+    TfLiteInterpreterOptionsSetNumThreads(Graph.TFLiteOps, 2);
+
+    // Store the loaded graph.
+    *GraphId = Env.NNGraph.size() - 1;
+    return static_cast<uint32_t>(WASINN::ErrNo::Success);
+    // -------------------------
```
Comment (q82419) [inline plugins/wasi_nn/wasinnfunc.cpp:333]: Delete the meaningless comment.

File: plugins/wasi_nn/wasinnfunc.cpp [hunk 6]
```diff
@@ -571,6 +640,50 @@ Expect<uint32_t> WasiNNSetInput::body(const Runtime::CallingFrame &Frame,
...
+  } else if (CxtRef.GraphRef.GraphBackend == WASINN::Backend::TensorflowLite) {
+#ifdef WASMEDGE_PLUGIN_WASI_NN_BACKEND_TFLITE
+    uint32_t InCnt = TfLiteInterpreterGetInputTensorCount(CxtRef.TFLiteInterp);
+    if (Index >= InCnt) {
+      spdlog::error("[WASI-NN] Invalid index id {} for the input, only {} "
+                    "inputs are allowed",
+                    Index, InCnt);
+      return static_cast<uint32_t>(WASINN::ErrNo::InvalidArgument);
+    }
+    uint32_t *Tensor = MemInst->getPointer<uint32_t *>(TensorPtr, 5);
```
Comment (q82419) [inline plugins/wasi_nn/wasinnfunc.cpp:653]: It seems like the `Tensor[2]` for the tensor type is not used?
Maybe we should give an additional check with the tflite API with the tensor type?

File: .github/workflows/build-extensions.yml [hunk 7]
```diff
@@ -223,9 +223,12 @@ jobs:
             system: manylinux2014 x86_64
             host_runner: ubuntu-latest
             docker_tag: manylinux2014_x86_64
+            test_path: test/plugins/wasi_nn
+            test_bin: wasiNNTests
             output_path: plugins/wasi_nn
             output_bin: libwasmedgePluginWasiNN.so
             build_type: Release
+            test: false
```
Comment (q82419) [inline .github/workflows/build-extensions.yml:231]: I think it's not necessary to test on the manylinux now.
We can refactor the CI in the future and add the test.

**Review by q82419 — CHANGES_REQUESTED** [review]
Seems great.
Please fix with the comments above and I'll test them in these days.

I've updated the `proposal/wasi_nn` branch. Please rebase on it to fix the MacOS CI failed.

File: .github/workflows/build-extensions.yml (hunk 7 above)
Comment (gusye1234) [reply .github/workflows/build-extensions.yml:231]: The test of PyTorch is only on `ubuntu`, only tflite tests on the `manylinux`

**Review by q82419 — APPROVED** [review]

**Pull request merged**